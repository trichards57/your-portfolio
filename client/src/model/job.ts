export type Outcome =
  | "DischargedOnScene"
  | "NotFound"
  | "StoodDown"
  | "Conveyed"
  | "Other";

export interface NewJob {
  age?: number;
  blueLights: boolean;
  category: 1 | 2 | 3 | 4 | 5;
  drove: boolean;
  gender?: "Male" | "Female";
  notes: string;
  outcome: Outcome;
  reflectionFlag: boolean;
  shift: string;
}

export interface JobSummary {
  age?: number;
  category: 1 | 2 | 3 | 4 | 5;
  gender?: "Male" | "Female";
  reflectionFlag: boolean;
}
