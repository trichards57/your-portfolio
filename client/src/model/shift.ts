export type RoleType = "EAC" | "AFA" | "CRU";

export interface Shift {
  crewMate?: string;
  date: string;
  duration: number;
  event: string;
  id: string;
  location?: string;
  role: RoleType;
}

export interface ShiftSummary extends Shift {
  loggedCalls: number;
}

export type NewShift = Omit<Shift, "id">;
