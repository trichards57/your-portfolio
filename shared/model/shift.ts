export interface ShiftSummary {
  id: number;
  date: string;
  event: string;
  location: string;
  duration: number;
  role: "EAC" | "AFA" | "CRU";
  crewMate: string;
  loggedCalls: number;
}
