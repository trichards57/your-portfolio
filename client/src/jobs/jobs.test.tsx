import { useAuth0 } from "@auth0/auth0-react";
import { render, waitFor } from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import React from "react";
import ReactDOM from "react-dom";
import { useHistory } from "react-router-dom";
import { JobsBase } from ".";
import { JobSummary } from "../model/job";

const testShiftId = "hijklmn";

jest.mock("@auth0/auth0-react");
jest.mock("react-router-dom", () => ({
  useLocation: () => ({
    search: `?shiftId=${testShiftId}`,
  }),
  useHistory: jest.fn(),
}));

const testToken = "abcdefg";

const endPoint = "/api/GetJobs";

const server = setupServer(
  rest.get(endPoint, (req, res, ctx) => {
    if (req.headers.get("authorization") === `Bearer ${testToken}`) {
      if (req.url.searchParams.get("shiftId") === testShiftId) {
        return res(
          ctx.json([
            {
              category: 1,
              id: "abcd",
              reflectionFlag: true,
              age: 10,
              gender: "Male",
            },
            {
              category: 2,
              id: "efgh",
              reflectionFlag: true,
              age: 10,
              gender: "Female",
            },
            {
              category: 3,
              id: "ijkl",
              reflectionFlag: false,
            },
          ] as JobSummary[])
        );
      }
      return res(ctx.status(404));
    }
    return res(ctx.status(401));
  })
);

beforeAll(() => server.listen());

beforeEach(() =>
  (useAuth0 as jest.Mock).mockReturnValue({
    getAccessTokenSilently: jest
      .fn()
      .mockReturnValue(Promise.resolve(testToken)),
  })
);

afterEach(() => server.resetHandlers());

afterAll(() => server.close());

it("renders without crashing", () => {
  const div = document.createElement("div");

  ReactDOM.render(<JobsBase />, div);

  ReactDOM.unmountComponentAtNode(div);
});

it("renders loading correctly", async () => {
  const res = render(<JobsBase />);

  expect(res.asFragment()).toMatchSnapshot();
});

it("renders successful load correctly", async () => {
  const res = render(<JobsBase />);

  await res.findByText("Male");

  expect(res.asFragment()).toMatchSnapshot();
});

it("redirects to the home page if not authorised", async () => {
  (useAuth0 as jest.Mock).mockReturnValue({
    getAccessTokenSilently: jest
      .fn()
      .mockReturnValue(Promise.resolve("BadToken")),
  });

  const testPush = jest.fn();

  (useHistory as jest.Mock).mockReturnValue({
    push: testPush,
  });

  render(<JobsBase />);

  await waitFor(() => expect(testPush).toBeCalledWith("/"));
});

it("displays an error if the request fails", async () => {
  server.use(rest.get(endPoint, (_, res, ctx) => res(ctx.status(500))));

  const res = render(<JobsBase />);

  await res.findByText("There was a problem speaking to the server.", {
    exact: false,
  });
});
