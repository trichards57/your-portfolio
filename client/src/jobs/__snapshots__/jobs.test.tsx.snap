// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders loading correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-16"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-18 mui-fixed MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          aria-label="open drawer"
          class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-19 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap"
        >
          Your Portfolio
        </h6>
      </div>
    </header>
    <nav
      aria-label="mailbox folders"
      class="makeStyles-drawer-17"
    />
    <main
      class="makeStyles-content-22"
    >
      <div
        class="makeStyles-toolbar-20"
      />
      <h2
        class="MuiTypography-root MuiTypography-h5"
      >
        Jobs
      </h2>
      <div
        class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <table
          class="MuiTable-root"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Job Number
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Age
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Gender
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              />
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                />
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
            </tr>
          </tbody>
        </table>
      </div>
    </main>
  </div>
</DocumentFragment>
`;

exports[`renders successful load correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-27"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-29 mui-fixed MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          aria-label="open drawer"
          class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-30 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap"
        >
          Your Portfolio
        </h6>
      </div>
    </header>
    <nav
      aria-label="mailbox folders"
      class="makeStyles-drawer-28"
    />
    <main
      class="makeStyles-content-33"
    >
      <div
        class="makeStyles-toolbar-31"
      />
      <h2
        class="MuiTypography-root MuiTypography-h5"
      >
        Jobs
      </h2>
      <div
        class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <table
          class="MuiTable-root"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Job Number
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Age
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Gender
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              />
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                10
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                Male
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                10
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                Female
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                3
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
              <td
                class="MuiTableCell-root MuiTableCell-body"
              />
            </tr>
          </tbody>
        </table>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
